tools/filetools.go

```diff
@@ -113,6 +115,60 @@ func CleanPaths(paths, delim string) (cleaned []string) {
 	return cleaned
 }
 
+var (
+	currentUser func() (*user.User, error)           = user.Current
+	lookupUser  func(who string) (*user.User, error) = user.Lookup
+)
+
+// ExpandPath returns a copy of path with any references to the current user's
+// home directory (spelled "~"), or a named user's home directory (spelled
+// "~user") in the path, sanitized to the calling filesystem's path separator
+// preference.
+//
+// If the "expand" argument is given as true, the resolved path to the named
+// user's home directory will expanded with filepath.EvalSymlinks.
+//
+// If either the current or named user does not have a home directory, an error
+// will be returned.
+//
+// Otherwise, the error returned will be nil, and the string returned will be
+// the expanded path.
+func ExpandPath(path string, expand bool) (string, error) {
+	if len(path) == 0 || path[0] != '~' {
+		return path, nil
+	}
+
+	var username string
+	if slash := strings.Index(path[1:], "/"); slash > -1 {
+		username = path[1 : slash+1]
+	} else {
+		username = path[1:]
+	}
+
+	var (
+		who *user.User
+		err error
+	)
+	if len(username) == 0 {
+		who, err = currentUser()
```

Comment:
There's a tricky edge case here that we haven't considered.  If we have just a tilde (e.g. `~/hooks`), we need to use the `$HOME` environment variable in preference to what the OS (e.g. `/etc/passwd`) tells us the user's home directory is.  This is Git's behavior (and the shell's) and I suspect people will want us to match it.

Reply:
That makes a lot of sense. I was initially mislead because even though I thought that Git's behavior was to use `$HOME` instead of `/etc/passwd` or similar, that that _is_ what Go's `user.Current()` was doing, because of [this][1] line.

But, I was puzzled since we do set `HOME="$TESTHOME"` before running each test's subshell. So, before sending this patch I tried to look into it myself, but couldn't figure out why the two disagreed.

Because of Go's conditional compilation, the line mentioned above wasn't what we were executing. Instead, we were running [this][2], which is in a different file (and uses `getpw()` instead of `os.Getenv("HOME")`).

So, I was able to re-wrap this function so that we effectively do:

```go
func getCurrentUser() (*user.User, error) {
        u, err := user.Current()
        if err != nil {
                return nil, err
        }
        u.HomeDir = os.Getenv("HOME")
        return u, nil
}
```

and call that, instead of refactoring the code to support one case that uses package `os/user`, and one that does not.

[1]: https://github.com/golang/go/blob/go1.11/src/os/user/lookup_stubs.go#L27
[2]: https://github.com/golang/go/blob/go1.11/src/os/user/cgo_lookup_unix.go#L31-L34